unnecessary panic log in hash join v2 when meet error during build stage and `needScanRowTableAfterProbeDone` is true

## Enhancement

[dump.txt](https://github.com/user-attachments/files/19676995/dump.txt)

For query
```
update /*+ hash_join_build(t9a420759) */ `t9a420759` set `t9a420759`.`colc80b` = '1971-07-14' ,`t9a420759`.`col29ab` = '5kw6a' ,`t9a420759`.`col9ecc` = 'q8' where `t9a420759`.`col414c` in ( select `colce0b` from `ta871c766` where `t9a420759`.`col9926` between 0 and 1 )  limit 21375803;
```
it will meet error
ERROR 1292 (22007): Truncated incorrect DOUBLE value: '9VfzBPaNbYL4eiG'
This is expected since incorrect data
But in TiDB's log, there is some panic related logs
```
[2025/04/10 10:08:42.305 +08:00] [ERROR] [wait_group_wrapper.go:198] ["panic in the wait group"] [recover="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1.1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:198\nruntime.gopanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:791\nruntime.panicmem\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:262\nruntime.sigpanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/signal_unix.go:917\ngithub.com/pingcap/tidb/pkg/executor/join.(*hashTableV2).totalRowCount\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_table_v2.go:243\ngithub.com/pingcap/tidb/pkg/executor/join.commonInitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/base_join_probe.go:712\ngithub.com/pingcap/tidb/pkg/executor/join.(*semiJoinProbe).InitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/semi_join_probe.go:40\ngithub.com/pingcap/tidb/pkg/executor/join.(*ProbeWorkerV2).scanRowTableAfterProbeDone\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:889\ngithub.com/pingcap/tidb/pkg/executor/join.(*HashJoinV2Exec).waitJoinWorkers.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:881\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:202"]
[2025/04/10 10:08:42.305 +08:00] [ERROR] [wait_group_wrapper.go:198] ["panic in the wait group"] [recover="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1.1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:198\nruntime.gopanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:791\nruntime.panicmem\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:262\nruntime.sigpanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/signal_unix.go:917\ngithub.com/pingcap/tidb/pkg/executor/join.(*hashTableV2).totalRowCount\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_table_v2.go:243\ngithub.com/pingcap/tidb/pkg/executor/join.commonInitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/base_join_probe.go:712\ngithub.com/pingcap/tidb/pkg/executor/join.(*semiJoinProbe).InitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/semi_join_probe.go:40\ngithub.com/pingcap/tidb/pkg/executor/join.(*ProbeWorkerV2).scanRowTableAfterProbeDone\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:889\ngithub.com/pingcap/tidb/pkg/executor/join.(*HashJoinV2Exec).waitJoinWorkers.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:881\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:202"]
[2025/04/10 10:08:42.305 +08:00] [ERROR] [wait_group_wrapper.go:198] ["panic in the wait group"] [recover="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1.1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:198\nruntime.gopanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:791\nruntime.panicmem\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:262\nruntime.sigpanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/signal_unix.go:917\ngithub.com/pingcap/tidb/pkg/executor/join.(*hashTableV2).totalRowCount\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_table_v2.go:243\ngithub.com/pingcap/tidb/pkg/executor/join.commonInitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/base_join_probe.go:712\ngithub.com/pingcap/tidb/pkg/executor/join.(*semiJoinProbe).InitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/semi_join_probe.go:40\ngithub.com/pingcap/tidb/pkg/executor/join.(*ProbeWorkerV2).scanRowTableAfterProbeDone\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:889\ngithub.com/pingcap/tidb/pkg/executor/join.(*HashJoinV2Exec).waitJoinWorkers.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:881\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:202"]
[2025/04/10 10:08:42.305 +08:00] [ERROR] [wait_group_wrapper.go:198] ["panic in the wait group"] [recover="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1.1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:198\nruntime.gopanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:791\nruntime.panicmem\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:262\nruntime.sigpanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/signal_unix.go:917\ngithub.com/pingcap/tidb/pkg/executor/join.(*hashTableV2).totalRowCount\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_table_v2.go:243\ngithub.com/pingcap/tidb/pkg/executor/join.commonInitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/base_join_probe.go:712\ngithub.com/pingcap/tidb/pkg/executor/join.(*semiJoinProbe).InitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/semi_join_probe.go:40\ngithub.com/pingcap/tidb/pkg/executor/join.(*ProbeWorkerV2).scanRowTableAfterProbeDone\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:889\ngithub.com/pingcap/tidb/pkg/executor/join.(*HashJoinV2Exec).waitJoinWorkers.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:881\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:202"]
[2025/04/10 10:08:42.305 +08:00] [ERROR] [wait_group_wrapper.go:198] ["panic in the wait group"] [recover="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1.1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:198\nruntime.gopanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:791\nruntime.panicmem\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/panic.go:262\nruntime.sigpanic\n\t/Users/feixu/.gvm/pkgsets/go1.23.6/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.8.darwin-arm64/src/runtime/signal_unix.go:917\ngithub.com/pingcap/tidb/pkg/executor/join.(*hashTableV2).totalRowCount\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_table_v2.go:243\ngithub.com/pingcap/tidb/pkg/executor/join.commonInitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/base_join_probe.go:712\ngithub.com/pingcap/tidb/pkg/executor/join.(*semiJoinProbe).InitForScanRowTable\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/semi_join_probe.go:40\ngithub.com/pingcap/tidb/pkg/executor/join.(*ProbeWorkerV2).scanRowTableAfterProbeDone\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:889\ngithub.com/pingcap/tidb/pkg/executor/join.(*HashJoinV2Exec).waitJoinWorkers.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/executor/join/hash_join_v2.go:881\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1\n\t/Users/feixu/dev/pingcap/tidb/pkg/util/wait_group_wrapper.go:202"]
```
This is not expected

--- a/pkg/executor/join/hash_join_base.go
+++ b/pkg/executor/join/hash_join_base.go
@@ -59,12 +59,12 @@ type hashJoinCtxBase struct {
 }
 
 type probeSideTupleFetcherBase struct {
-	ProbeSideExec      exec.Executor
-	probeChkResourceCh chan *probeChkResource
-	probeResultChs     []chan *chunk.Chunk
-	requiredRows       int64
-	joinResultChannel  chan *hashjoinWorkerResult
-	probeFetchSkipped  bool
+	ProbeSideExec       exec.Executor
+	probeChkResourceCh  chan *probeChkResource
+	probeResultChs      []chan *chunk.Chunk
+	requiredRows        int64
+	joinResultChannel   chan *hashjoinWorkerResult
+	canSkipScanRowTable bool
 }
 
 func (fetcher *probeSideTupleFetcherBase) initializeForProbeBase(concurrency uint, joinResultChannel chan *hashjoinWorkerResult) {
@@ -99,18 +99,21 @@ func (fetcher *probeSideTupleFetcherBase) handleProbeSideFetcherPanic(r any) {
 type isBuildSideEmpty func() bool
 type isSpillTriggered func() bool
 
-func wait4BuildSide(isBuildEmpty isBuildSideEmpty, checkSpill isSpillTriggered, canSkipIfBuildEmpty, needScanAfterProbeDone bool, hashJoinCtx *hashJoinCtxBase) (skipProbe bool) {
+func wait4BuildSide(isBuildEmpty isBuildSideEmpty, checkSpill isSpillTriggered, canSkipIfBuildEmpty, needScanAfterProbeDone bool, hashJoinCtx *hashJoinCtxBase) (skipProbe bool, skipScanRowTable bool) {
 	var err error
 	skipProbe = false
+	skipScanRowTable = false
 	buildFinishes := false
 	select {
 	case <-hashJoinCtx.closeCh:
 		// current executor is closed, no need to probe
 		skipProbe = true
+		skipScanRowTable = true
 	case err = <-hashJoinCtx.buildFinished:
 		if err != nil {
 			// build meet error, no need to probe
 			skipProbe = true
+			skipScanRowTable = true
 		} else {
 			buildFinishes = true
 		}
@@ -131,7 +134,7 @@ func wait4BuildSide(isBuildEmpty isBuildSideEmpty, checkSpill isSpillTriggered, 
 			hashJoinCtx.finished.Store(true)
 		}
 	}
-	return skipProbe
+	return skipProbe, skipScanRowTable
 }
 
 func (fetcher *probeSideTupleFetcherBase) getProbeSideResource(shouldLimitProbeFetchSize bool, maxChunkSize int, hashJoinCtx *hashJoinCtxBase) *probeChkResource {
@@ -189,10 +192,12 @@ func (fetcher *probeSideTupleFetcherBase) fetchProbeSideChunks(ctx context.Conte
 					probeSideResult.Reset()
 				}
 			})
-			skipProbe := wait4BuildSide(isBuildEmpty, checkSpill, canSkipIfBuildEmpty, needScanAfterProbeDone, hashJoinCtx)
+			skipProbe, skipScanRowTable := wait4BuildSide(isBuildEmpty, checkSpill, canSkipIfBuildEmpty, needScanAfterProbeDone, hashJoinCtx)
+			if skipScanRowTable {
+				fetcher.canSkipScanRowTable = true
+			}
 			if skipProbe {
 				// there is no need to probe, so just return
-				fetcher.probeFetchSkipped = true
 				return
 			}
 			hasWaitedForBuild = true

--- a/pkg/executor/join/hash_join_base.go
+++ b/pkg/executor/join/hash_join_base.go
@@ -64,6 +64,7 @@ type probeSideTupleFetcherBase struct {
 	probeResultChs     []chan *chunk.Chunk
 	requiredRows       int64
 	joinResultChannel  chan *hashjoinWorkerResult
+	probeFetchSkipped  bool
 }
 
 func (fetcher *probeSideTupleFetcherBase) initializeForProbeBase(concurrency uint, joinResultChannel chan *hashjoinWorkerResult) {
@@ -191,6 +192,7 @@ func (fetcher *probeSideTupleFetcherBase) fetchProbeSideChunks(ctx context.Conte
 			skipProbe := wait4BuildSide(isBuildEmpty, checkSpill, canSkipIfBuildEmpty, needScanAfterProbeDone, hashJoinCtx)
 			if skipProbe {
 				// there is no need to probe, so just return
+				fetcher.probeFetchSkipped = true
 				return
 			}
 			hasWaitedForBuild = true

--- a/pkg/executor/join/hash_join_v2.go
+++ b/pkg/executor/join/hash_join_v2.go
@@ -878,7 +878,7 @@ func (e *HashJoinV2Exec) waitJoinWorkers(start time.Time) {
 	if !e.ProbeSideTupleFetcher.probeFetchSkipped {
 		// only scan the row table if probe fetch is not skipped
 		// probe fetch is skipped if
-		// 1. there is some error during build/probe stage
+		// 1. there is some error during build stage
 		// 2. the hash table is empty
 		// in both cases, there is no need to scan the row table again
 		if e.ProbeWorkers[0] != nil && e.ProbeWorkers[0].JoinProbe.NeedScanRowTable() {

--- a/pkg/executor/join/hash_join_v2.go
+++ b/pkg/executor/join/hash_join_v2.go
@@ -787,7 +787,7 @@ func (e *HashJoinV2Exec) initializeForProbe() {
 	e.joinResultCh = make(chan *hashjoinWorkerResult, e.Concurrency+1)
 	e.ProbeSideTupleFetcher.initializeForProbeBase(e.Concurrency, e.joinResultCh)
 	e.ProbeSideTupleFetcher.canSkipProbeIfHashTableIsEmpty = e.canSkipProbeIfHashTableIsEmpty()
-	e.ProbeSideTupleFetcher.probeFetchSkipped = false
+	e.ProbeSideTupleFetcher.canSkipScanRowTable = false
 
 	for i := uint(0); i < e.Concurrency; i++ {
 		e.ProbeWorkers[i].initializeForProbe(e.ProbeSideTupleFetcher.probeChkResourceCh, e.ProbeSideTupleFetcher.probeResultChs[i], e)
@@ -875,12 +875,7 @@ func (e *HashJoinV2Exec) waitJoinWorkers(start time.Time) {
 		}
 	}
 
-	if !e.ProbeSideTupleFetcher.probeFetchSkipped {
-		// only scan the row table if probe fetch is not skipped
-		// probe fetch is skipped if
-		// 1. there is some error during build stage
-		// 2. the hash table is empty
-		// in both cases, there is no need to scan the row table again
+	if !e.ProbeSideTupleFetcher.canSkipScanRowTable {
 		if e.ProbeWorkers[0] != nil && e.ProbeWorkers[0].JoinProbe.NeedScanRowTable() {
 			for i := uint(0); i < e.Concurrency; i++ {
 				var workerID = i

--- a/pkg/executor/join/hash_join_v2.go
+++ b/pkg/executor/join/hash_join_v2.go
@@ -787,6 +787,7 @@ func (e *HashJoinV2Exec) initializeForProbe() {
 	e.joinResultCh = make(chan *hashjoinWorkerResult, e.Concurrency+1)
 	e.ProbeSideTupleFetcher.initializeForProbeBase(e.Concurrency, e.joinResultCh)
 	e.ProbeSideTupleFetcher.canSkipProbeIfHashTableIsEmpty = e.canSkipProbeIfHashTableIsEmpty()
+	e.ProbeSideTupleFetcher.probeFetchSkipped = false
 
 	for i := uint(0); i < e.Concurrency; i++ {
 		e.ProbeWorkers[i].initializeForProbe(e.ProbeSideTupleFetcher.probeChkResourceCh, e.ProbeSideTupleFetcher.probeResultChs[i], e)
@@ -874,14 +875,21 @@ func (e *HashJoinV2Exec) waitJoinWorkers(start time.Time) {
 		}
 	}
 
-	if e.ProbeWorkers[0] != nil && e.ProbeWorkers[0].JoinProbe.NeedScanRowTable() {
-		for i := uint(0); i < e.Concurrency; i++ {
-			var workerID = i
-			e.workerWg.RunWithRecover(func() {
-				e.ProbeWorkers[workerID].scanRowTableAfterProbeDone()
-			}, e.handleJoinWorkerPanic)
+	if !e.ProbeSideTupleFetcher.probeFetchSkipped {
+		// only scan the row table if probe fetch is not skipped
+		// probe fetch is skipped if
+		// 1. there is some error during build/probe stage
+		// 2. the hash table is empty
+		// in both cases, there is no need to scan the row table again
+		if e.ProbeWorkers[0] != nil && e.ProbeWorkers[0].JoinProbe.NeedScanRowTable() {
+			for i := uint(0); i < e.Concurrency; i++ {
+				var workerID = i
+				e.workerWg.RunWithRecover(func() {
+					e.ProbeWorkers[workerID].scanRowTableAfterProbeDone()
+				}, e.handleJoinWorkerPanic)
+			}
+			e.workerWg.Wait()
 		}
-		e.workerWg.Wait()
 	}
 }
 

